addons/sale_purchase/models/sale_order_line.py

```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
+        return purchase_order
+
     def _purchase_service_create(self, quantity=False):
         """ On Sales Order confirmation, some lines (services ones) can create a purchase order line and maybe a purchase order.
             If a line should create a RFQ, it will check for existing PO. If no one is find, the SO line will create one, then adds
             a new PO line. The created purchase order line will be linked to the SO line.
             :param quantity: the quantity to force on the PO line, expressed in SO line UoM
         """
-        PurchaseOrder = self.env['purchase.order']
         supplier_po_map = {}
-        sale_line_purchase_map = {}
+        purchase_lines = self.env['purchase.order.line']
         for line in self:
-            line = line.with_company(line.company_id)
-            # determine vendor of the order (take the first matching company and product)
-            suppliers = line.product_id._select_seller(quantity=line.product_uom_qty, uom_id=line.product_uom)
-            if not suppliers:
-                raise UserError(_("There is no vendor associated to the product %s. Please define a vendor for this product.") % (line.product_id.display_name,))
-            supplierinfo = suppliers[0]
+            line = line.with_company(line._purchase_service_get_company())
+            supplierinfo = self._purchase_service_match_supplier()
             partner_supplier = supplierinfo.partner_id
 
             # determine (or create) PO
             purchase_order = supplier_po_map.get(partner_supplier.id)
             if not purchase_order:
-                purchase_order = PurchaseOrder.search([
-                    ('partner_id', '=', partner_supplier.id),
-                    ('state', '=', 'draft'),
-                    ('company_id', '=', line.company_id.id),
-                ], limit=1)
-            if not purchase_order:
-                values = line._purchase_service_prepare_order_values(supplierinfo)
-                purchase_order = PurchaseOrder.with_context(mail_create_nosubscribe=True).create(values)
-            else:  # update origin of existing PO
+                purchase_order = self._match_or_create_purchase_order(supplierinfo)
+            else:  # if not already updated origin in this loop
                 so_name = line.order_id.name
-                origins = []
-                if purchase_order.origin:
-                    origins = purchase_order.origin.split(', ') + origins
+                origins = (purchase_order.origin or '').split(', ')
                 if so_name not in origins:
-                    origins += [so_name]
-                    purchase_order.write({
-                        'origin': ', '.join(origins)
-                    })
-            supplier_po_map[partner_supplier.id] = purchase_order
+                    purchase_order.write({'origin': ', '.join(origins + [so_name])})
+                supplier_po_map[partner_supplier.id] = purchase_order
```

Comment:
If the PO was created on line 249, it should still be added to the map, so we need to put it outside of the if/else block (as in the extracted code):
```suggestion
            supplier_po_map[partner_supplier.id] = purchase_order
```

Reply:
not sure, if we create it, we already know that we have the origin on the PO (purpose of the else part).

So the second time, the match will return the correct purchase order, and avoid to resplit/join the origin without modification :thinking: but `match` vs `split/join` not sure what is the worst. All this code is not optimized and could be rewrite in master.
I will unindent it in stable, since it was not obvious ...